README.md
## Introduction

flutter_hl_vnpay is a VNPay plugin for [Flutter](https://flutter.dev), a new app SDK to help developers build modern multi-platform apps.

## Getting Started
#### **iOS**
In the **ios/Runner/Info.plist** let’s add:

```dart
	<dict>
	     <key>NSAppTransportSecurity</key>
          <dict>
                <key>NSAllowsArbitraryLoads</key>
                <true/>
                <key>NSAllowsArbitraryLoadsInWebContent</key>
                <true/>
          </dict>
```

```dart
   	<key>CFBundleURLTypes</key>
   	<array>
   		<dict>
   			<key>CFBundleTypeRole</key>
   			<string>Editor</string>
   			<key>CFBundleURLName</key>
   			<string>com.hlsolutions.hoishipper</string>
   			<key>CFBundleURLSchemes</key>
   			<array>
   				<string>hlsolutions</string>
   			</array>
   		</dict>
   	</array>

```

#### **Android**

## Usage
#### 1\. Depend
Add this to you package's `pubspec.yaml` file:

```yaml
dependencies:
  flutter_hl_vnpay: ^2.0.1
  crypto: 2.1.5
```
#### 2\. Install
Run command:

```bash
$ flutter pub get
```

#### 3\. Import

Import in Dart code:

```dart
import 'package:flutter/services.dart';
import 'package:flutter_hl_vnpay/flutter_hl_vnpay.dart';
import 'package:crypto/crypto.dart';
```

#### 4\. Using VNPAY
https://sandbox.vnpayment.vn/apis/docs/huong-dan-tich-hop/#t%E1%BA%A1o-url-thanh-to%C3%A1n

```dart
try {
      String url = 'http://sandbox.vnpayment.vn/paymentv2/vpcpay.html';
      String tmnCode = 'XXX'; // Get from VNPay
      String hashKey = 'XXX'; // Get from VNPay

      final params = <String, dynamic>{
        'vnp_Command': 'pay',
        'vnp_Amount': '3000000',
        'vnp_CreateDate': '20210315151908',
        'vnp_CurrCode': 'VND',
        'vnp_IpAddr': '192.168.15.102',
        'vnp_Locale': 'vn',
        'vnp_OrderInfo': '5fa66b8b5f376a000417e501 pay coin 30000 VND',
        'vnp_ReturnUrl': 'XXX', // Your Server https://sandbox.vnpayment.vn/apis/docs/huong-dan-tich-hop/#code-returnurl
        'vnp_TmnCode': tmnCode,
        'vnp_TxnRef': '604f187c862cd70004478e',
        'vnp_Version': '2.0.0'
      };

      final sortedParams = FlutterHlVnpay.instance.sortParams(params);
      final hashDataBuffer = new StringBuffer();
      sortedParams.forEach((key, value) {
        hashDataBuffer.write(key);
        hashDataBuffer.write('=');
        hashDataBuffer.write(value);
        hashDataBuffer.write('&');
      });
      final hashData = hashDataBuffer.toString().substring(0, hashDataBuffer.length - 1);
      final query = Uri(queryParameters: sortedParams).query;
      print('hashData = $hashData');
      print('query = $query');

      var bytes = utf8.encode(hashKey + hashData.toString());
      final vnpSecureHash = sha256.convert(bytes);
      final paymentUrl = "$url?$query&vnp_SecureHashType=SHA256&vnp_SecureHash=$vnpSecureHash";
      print('paymentUrl = $paymentUrl');
      paymentResultCodeCode = (await FlutterHlVnpay.instance.show(paymentUrl: paymentUrl, tmnCode: tmnCode)).toString();
    } on PlatformException {
      paymentResultCodeCode = 'Failed to get payment result code';
    }
```

CHANGELOG.md
## 2.0.1
Mirgate VNP android lib

## 2.0.0
Migrate null safe
Config schemes to close vnpay screen

## 1.0.1
Release iOS, Android


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_hl_vnpay_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/flutter_hl_vnpay.dart

import 'dart:async';

import 'package:flutter/services.dart';

class FlutterHlVnpay {
  final MethodChannel _channel = const MethodChannel('flutter_hl_vnpay');
  final StreamController<MethodCall> _methodStreamController = new StreamController.broadcast();
  Stream<MethodCall> get _methodStream => _methodStreamController.stream;

  FlutterHlVnpay._() {
    _channel.setMethodCallHandler((MethodCall call) async {
      print('[setMethodCallHandler] call = $call');
      _methodStreamController.add(call);
    });
  }

  static FlutterHlVnpay _instance = new FlutterHlVnpay._();
  static FlutterHlVnpay get instance => _instance;

  Future<int?> show({
    required String paymentUrl,
    required String tmnCode,
    String scheme = '',
    bool isSandbox = true,
    String backAlert = 'Bạn có chắc chắn trở lại ko?',
    String title = 'Nạp tiền qua VNPay',
    String iconBackName = 'AppIcon',
    String beginColor = '#00B14F',
    String endColor = '#00B14F',
    String titleColor = '#FFFFFF',
  }) async {

    final params = <String, dynamic>{
      "scheme": scheme,
      "isSandbox": isSandbox,
      "paymentUrl": paymentUrl,
      "tmn_code": tmnCode,
      "backAlert": backAlert,
      "title": title,
      "iconBackName": iconBackName,
      "beginColor": beginColor,
      "endColor": endColor,
      "titleColor": titleColor,
    };
    await _channel.invokeMethod('show', params);
    await for (MethodCall m in _methodStream) {
      if (m.method == "PaymentBack") {
        return m.arguments['resultCode'] as int?;
      }
    }
  }

  Map<String, dynamic> sortParams(Map<String, dynamic> params) {
    final sortedParams = <String, dynamic>{};
    final keys = params.keys.toList()..sort();
    for (String key in keys) {
      sortedParams[key] = params[key];
    }
    return sortedParams;
  }
}


ios/Classes/FlutterHlVnpayPlugin.m
#import "FlutterHlVnpayPlugin.h"
#import <CallAppSDK/CallAppInterface.h>

@interface FlutterHlVnpayPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@property(nonatomic, copy) NSString *latestScheme;
@end

@implementation FlutterHlVnpayPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutter_hl_vnpay"
                                     binaryMessenger:[registrar messenger]];
    FlutterHlVnpayPlugin* instance = [[FlutterHlVnpayPlugin alloc] init];
    instance.channel = channel;
    [registrar addMethodCallDelegate:instance channel:channel];
    [registrar addApplicationDelegate:instance];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"show" isEqualToString:call.method]) {
        [self handleShow:call];
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)handleShow:(FlutterMethodCall*)call {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sdkAction:)
                                                 name:@"SDK_COMPLETED" object:nil];
    [CallAppInterface setHomeViewController:[self viewControllerWithWindow:nil]];

    NSDictionary *value = [call arguments];
    bool isSandbox = value[@"isSandbox"];
    NSString *scheme = value[@"scheme"];
    NSString *backAlert = value[@"backAlert"];
    NSString *paymentUrl = value[@"paymentUrl"];
    NSString *title = value[@"title"];
    NSString *iconBackName = value[@"iconBackName"];
    NSString *beginColor = value[@"beginColor"];
    NSString *endColor = value[@"endColor"];
    NSString *titleColor = value[@"titleColor"];
    NSString *tmn_code = value[@"tmn_code"];
    
    self.latestScheme = scheme;

    [CallAppInterface setSchemes:scheme];
    [CallAppInterface setIsSandbox:isSandbox];
    [CallAppInterface setAppBackAlert:backAlert];
    [CallAppInterface showPushPaymentwithPaymentURL:paymentUrl
                                          withTitle:title
                                       iconBackName:iconBackName
                                         beginColor:beginColor
                                           endColor:endColor
                                         titleColor:titleColor
                                           tmn_code:tmn_code];
}

-(void)sdkAction:(NSNotification*)notification{
    if([notification.name isEqualToString:@"SDK_COMPLETED"]){
        [[NSNotificationCenter defaultCenter] removeObserver:self];

        NSString *actionValue=[notification.object valueForKey:@"Action"];
        if ([@"AppBackAction" isEqualToString:actionValue]) {
            //Người dùng nhấn back từ sdk để quay lại
            [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@-1}];
            return;
        }
        if ([@"CallMobileBankingApp" isEqualToString:actionValue]) {
        //Người dùng nhấn chọn thanh toán qua app thanh toán (Mobile Banking, Ví...)
             [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@10}];
             return;
        }
        if ([@"WebBackAction" isEqualToString:actionValue]) {
            //Người dùng nhấn back từ trang thanh toán thành công khi thanh toán qua thẻ khi gọi đến http://sdk.merchantbackapp
            [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@24}];
            return;
        }
        if ([@"FaildBackAction" isEqualToString:actionValue]) {
             [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@99}];
             return;
        }
        if ([@"FailBackAction" isEqualToString:actionValue]) {
            [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@99}];
            return;
        }
        if ([@"SuccessBackAction" isEqualToString:actionValue]) {
            [_channel invokeMethod:@"PaymentBack" arguments:@{@"resultCode":@0}];
            return;
        }
    }
}

- (UIViewController *)viewControllerWithWindow:(UIWindow *)window {
    UIWindow *windowToUse = window;
    if (windowToUse == nil) {
        for (UIWindow *window in [UIApplication sharedApplication].windows) {
            if (window.isKeyWindow) {
                windowToUse = window;
                break;
            }
        }
    }

    UIViewController *topController = windowToUse.rootViewController;
    while (topController.presentedViewController) {
        topController = topController.presentedViewController;
    }
    return topController;
}

- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options {
    NSString *latestLink = [url absoluteString];
    NSString *scheme = [url scheme];
    NSString *host = [url host];
    if ([@"vnpay" isEqualToString:host] && [self.latestScheme isEqualToString:scheme]) {
        UIViewController *topController = [self viewControllerWithWindow:nil];
        UIWindow *windowToUse = nil;
        if (windowToUse == nil) {
            for (UIWindow *window in [UIApplication sharedApplication].windows) {
                if (window.isKeyWindow) {
                    windowToUse = window;
                    break;
                }
            }
        }
       
        if ([topController isKindOfClass:[FlutterViewController class]] == false) {
            [topController dismissViewControllerAnimated:YES completion:nil];
        }
    }
    return YES;
}

@end


ios/Frameworks/CallAppSDK.framework/Headers/CallAppInterface.h
//
//  CallAppInterface.h
//  CallAppSDK
//
//  Created by Tien Nguyen on 3/18/19.
//  Copyright © 2019 Tien Nguyen. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

NS_ASSUME_NONNULL_BEGIN

@interface CallAppInterface : NSObject

+(void)setAppNotAvailableAlert:(NSString*)alert;//Default: @"Thiết bị chưa cài đặt ứng dụng thanh toán này, bạn có muốn cài đặt không?"
+ (NSString*)getAppNotAvailableAlert;

+(void)setAppBackAlert:(NSString*)alert;//Default: @"Bạn có chắc chắn muốn trở lại không?"
+ (NSString*)getAppBackAlert;

+(void)setHomeViewController:(UIViewController*)vc;

//+(void)showPushPaymentVC:(UIViewController*)vc withPaymentURL:(NSString*)paymentStr withTitle:(NSString*)titleStr iconBackName:(NSString*)backIcon beginColor:(NSString*)beginColor endColor:(NSString*)endColor titleColor:(NSString*)titleColor tmn_code:(NSString*)tmn_code;

+(void)showPushPaymentwithPaymentURL:(NSString*)paymentStr withTitle:(NSString*)titleStr iconBackName:(NSString*)backIcon beginColor:(NSString*)beginColor endColor:(NSString*)endColor titleColor:(NSString*)titleColor tmn_code:(NSString*)tmn_code;

+ (void)setSchemes:(NSString*)schm;
+ (void)setIsSandbox:(BOOL)isSandBox;

@end


/**
 - Các tham số:
 1. URL: lấy từ hệ thống VD: https://sandbox.vnpayment.vn/tryitnow/Home/CreateOrder
 2. Title của màn hình chọn phương thức thanh toán: VD: "Thanh toán"
 3. iconBackName của màn hình thanh toán:  VD: ion_back
 4: beginColor: Màu bắt đầu của header:  VD: F06744
 5: endColor: Màu kết thúc của header:  VD: E26F2C
 6. tmn_code: VD: 2QXUI4J4
 7: titleColor: Màu của header title
 +(void)setAppBackAlert:(NSString*)alert;//Default: @"Bạn có chắc chắn muốn trở lại không?"
 +(void)setAppNotAvailableAlert:(NSString*)alert;//Default: @"Thiết bị chưa cài đặt ứng dụng thanh toán này, bạn có muốn cài đặt không?"
 7. [CallAppInterface setSchemes:@"sampleApp"];  Hàm set schemes của app để ứng dụng MB gọi quay trở lại khi thanh toán xong. Tham số này set trong URL Types của ứng dụng
 
 
 [CallAppInterface setSchemes:@"sampleapp"];
 [CallAppInterface setAppBackAlert:@"Bạn có chắc chắn trở lại ko?"];
 
 [CallAppInterface showPushPaymentVC:self withPaymentURL:@"https://sandbox.vnpayment.vn/tryitnow/Home/CreateOrder" withTitle:@"Thanh toans" iconBackName:@"ion_back" beginColor:@"F06744" endColor:@"E26F2C" titleColor:@"ffffff" tmn_code:@"2QXUI4J4"];
 
 
 */




NS_ASSUME_NONNULL_END


android/src/main/kotlin/com/hlsolutions/flutter_hl_vnpay/flutter_hl_vnpay/FlutterHlVnpayPlugin.kt
package com.hlsolutions.flutter_hl_vnpay.flutter_hl_vnpay

import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.util.Log
import androidx.annotation.NonNull
import com.vnpay.authentication.VNP_AuthenticationActivity

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/** FlutterHlVnpayPlugin */
class FlutterHlVnpayPlugin : FlutterPlugin, ActivityAware, PluginRegistry.ActivityResultListener, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var activityBinding: ActivityPluginBinding? = null
    private var flutterPluginBinding: FlutterPlugin.FlutterPluginBinding? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterPluginBinding = flutterPluginBinding
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_hl_vnpay")
        channel.setMethodCallHandler(this)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.addActivityResultListener(this)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeActivityResultListener(this)
        activityBinding = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterPluginBinding = binding
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        if (call.method == "show") {
            this.handleShow(call)
            result.success(null)
        } else {
            result.notImplemented()
        }
    }

    private fun handleShow(@NonNull call: MethodCall) {
        val params = call.arguments as HashMap<*, *>
        val paymentUrl = params["paymentUrl"] as String
        val scheme = params["scheme"] as String
        val tmnCode = params["tmn_code"] as String
        val intent = Intent(flutterPluginBinding!!.applicationContext, VNP_AuthenticationActivity::class.java).apply {
            putExtra("url", paymentUrl)
            putExtra("scheme", scheme)
            putExtra("tmn_code", tmnCode)
        }
        VNP_AuthenticationActivity.setSdkCompletedCallback { action ->
            Log.wtf("VNP_AuthenticationActivity", "action: $action")
            if (action == "AppBackAction") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to -1))
            }
            if (action == "CallMobileBankingApp") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to 10))
            }
            if (action == "WebBackAction") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to 24))
            }
            if (action == "FaildBackAction") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to 99))
            }
            if (action == "FailBackAction") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to 99))
            }
            if (action == "SuccessBackAction") {
                channel.invokeMethod("PaymentBack", hashMapOf("resultCode" to 0))
            }

            //action == AppBackAction
            //Người dùng nhấn back từ sdk để quay lại

            //action == CallMobileBankingApp
            //Người dùng nhấn chọn thanh toán qua app thanh toán (Mobile Banking, Ví...)
            //lúc này app tích hợp sẽ cần lưu lại mã giao dịch thanh toán (vnp_TxnRef). Khi người dùng mở lại app tích hợp với cheme thì sẽ gọi kiểm tra trạng thái thanh toán của mã TxnRef đó kiểm tra xem đã thanh toán hay chưa để thực hiện nghiệp vụ kết thúc thanh toán / thông báo kết quả cho khách hàng..

            //action == WebBackAction
            //Tạo nút sự kiện cho user click từ return url của merchant chuyển hướng về URL: http://cancel.sdk.merchantbackapp
            // vnp_ResponseCode == 24 / Khách hàng hủy thanh toán.

            //action == FaildBackAction
            //Tạo nút sự kiện cho user click từ return url của merchant chuyển hướng về URL: http://fail.sdk.merchantbackapp
            // vnp_ResponseCode != 00 / Giao dịch thanh toán không thành công

            //action == SuccessBackAction
            //Tạo nút sự kiện cho user click từ return url của merchant chuyển hướng về URL: http://success.sdk.merchantbackapp
            //vnp_ResponseCode == 00) / Giao dịch thành công
        }
        activityBinding?.activity?.startActivity(intent)
//        activityBinding?.activity?.startActivityForResult(intent, 99)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
      if (requestCode == 99) {
        channel.invokeMethod("resultCode", resultCode)
      }
      return false
    }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:convert';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_hl_vnpay/flutter_hl_vnpay.dart';
import 'package:crypto/crypto.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _paymentResultCodeCode = 'Unknown';

  @override
  void initState() {
    super.initState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
   _onBuyCoinPressed() async {
    String paymentResultCodeCode;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      String url = 'http://sandbox.vnpayment.vn/paymentv2/vpcpay.html';
      String tmnCode = 'XXX'; // Get from VNPay
      String hashKey = 'XXX'; // Get from VNPay

      final params = <String, dynamic>{
        'vnp_Command': 'pay',
        'vnp_Amount': '3000000',
        'vnp_CreateDate': '20210315151908',
        'vnp_CurrCode': 'VND',
        'vnp_IpAddr': '192.168.15.102',
        'vnp_Locale': 'vn',
        'vnp_OrderInfo': '5fa66b8b5f376a000417e501 pay coin 30000 VND',
        'vnp_ReturnUrl': 'XXX', // Your Server https://sandbox.vnpayment.vn/apis/docs/huong-dan-tich-hop/#code-returnurl
        'vnp_TmnCode': tmnCode,
        'vnp_TxnRef': DateTime
            .now()
            .millisecondsSinceEpoch.toString(),
        'vnp_Version': '2.0.0'
      };

      final sortedParams = FlutterHlVnpay.instance.sortParams(params);
      final hashDataBuffer = new StringBuffer();
      sortedParams.forEach((key, value) {
        hashDataBuffer.write(key);
        hashDataBuffer.write('=');
        hashDataBuffer.write(value);
        hashDataBuffer.write('&');
      });
      final hashData = hashDataBuffer.toString().substring(0, hashDataBuffer.length - 1);
      final query = Uri(queryParameters: sortedParams).query;
      print('hashData = $hashData');
      print('query = $query');

      var bytes = utf8.encode(hashKey + hashData.toString());
      final vnpSecureHash = sha256.convert(bytes);
      final paymentUrl = "$url?$query&vnp_SecureHashType=SHA256&vnp_SecureHash=$vnpSecureHash";
      print('paymentUrl = $paymentUrl');
      paymentResultCodeCode = (await FlutterHlVnpay.instance.show(paymentUrl: paymentUrl, tmnCode: tmnCode, scheme: 'hlsolutions')).toString();
    } on PlatformException {
      paymentResultCodeCode = 'Failed to get payment result code';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _paymentResultCodeCode = paymentResultCodeCode;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            children: [
              Text('Result Code: $_paymentResultCodeCode\n'),
              RaisedButton(
                onPressed: this._onBuyCoinPressed,
                child: Text('10.000 VND'),
              )
            ],
          ),
        ),
      ),
    );
  }
}
